File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>>ans;
    void solve(vector<int>&arr, int target, int cur, vector<int>&temp) {
        if(target == 0) {
            ans.push_back(temp);
            return;
        }
        if(target < 0 || cur >= arr.size()) return;
        unordered_set<int>st;
        for(int i=cur; i<arr.size(); i++) {
            if(st.find(arr[i]) != st.end()) continue;
            st.insert(arr[i]);
            temp.push_back(arr[i]);
            solve(arr, target - arr[i], i+1, temp);
            temp.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int>temp;
        sort(candidates.begin(), candidates.end());
        solve(candidates, target, 0, temp);
        return ans;
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        solve(candidates, target, new ArrayList<>(), 0);
        return ans;
    }
    private void solve(int[] arr, int target, List<Integer>temp, int cur) {
        if(target == 0) {
            ans.add(new ArrayList<>(temp));
            return;
        }
        if(target < 0 || cur >= arr.length) return;
        Set<Integer>st = new HashSet<>();
        for(int i=cur; i<arr.length; i++) {
            if(st.contains(arr[i])) continue;
            st.add(arr[i]);
            temp.add(arr[i]);
            solve(arr, target - arr[i], temp, i+1);
            temp.remove(temp.size()-1);
        }
    }
}